### PR TITLE
Add `urllib3.contrib.socks`; improve `urllib3.connectionpool`

### PR DESCRIPTION
**Pull request by kkirsche**
This is required to address some re-exports in https://github.com/python/typeshed/pull/8439

The source code is available here:
https://github.com/urllib3/urllib3/blob/main/src/urllib3/contrib/socks.py

### CODE REVIEW

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (srittau) [inline stubs/urllib3/urllib3/contrib/socks.pyi:9]: Why not `_TYPE_SOCKS_OPTIONS` like in the implementation?

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (srittau) [inline stubs/urllib3/urllib3/contrib/socks.pyi:18]: As we might reasonably be able to type `args` and `kwargs`:

```suggestion
    def __init__(self, _socks_options: _SockOptions, *args, **kwargs) -> None: ...
```

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (srittau) [inline stubs/urllib3/urllib3/contrib/socks.pyi:40]: Same here:

```suggestion
        **connection_pool_kw,
```

**Review by srittau — DISMISSED** [review]

**Comment by AlexWaygood** [comment]
As well as @srittau's points, note that this module only imports successfully if the optional dependency `PySocks` is installed (according to https://github.com/urllib3/urllib3/blob/16d2976041c5b8b0fe5f790e8edc9c50f318ee7b/src/urllib3/contrib/socks.py#L50-L52). You'll need the optional dependency to a `requirements-stubtest.txt` file in order for stubtest to be able to import it in CI (see https://github.com/python/typeshed/blob/master/stubs/tqdm/%40tests/requirements-stubtest.txt for an example).

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (kkirsche) [reply stubs/urllib3/urllib3/contrib/socks.pyi]: Updated. How would I type hint these given that they are `args` / `kwargs` of the `HTTPConnection` as seen in this?

https://github.com/urllib3/urllib3/blob/1.26.x/src/urllib3/connection.py#L115

With type hints in 2.x branch (missing the strict parameter):
https://github.com/urllib3/urllib3/blob/main/src/urllib3/connection.py#L136-L148

File: stubs/urllib3/@tests/requirements-stubtest.txt
Comment (AlexWaygood) [inline stubs/urllib3/@tests/requirements-stubtest.txt:1]: Does this really need to be so specific? (Honest question -- I haven't checked, and I don't know much about `PySocks`!)

Comment (kkirsche) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: https://github.com/urllib3/urllib3/blob/1.26.x/setup.py#L99

Comment (AlexWaygood) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: Thanks for the reference! It'll be annoying if we have to continually update this to stop our CI from failing, but I guess if that's what the runtime requires, we should probably have the same requirement here.

Comment (kkirsche) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: Does stubtest support the concept of `extras`? The easy way to solve this would be to load a list of extras to install with the package when doing tests

Comment (AlexWaygood) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: > Does stubtest support the concept of `extras`? The easy way to solve this would be to load a list of extras to install with the package when doing tests

Yeah, that's basically the purpose of the `stubtest-requirements.txt` files. All required dependencies are automatically installed by stubtest prior to it stubtesting; this file is for any additional dependencies that we want to specify (which are generally the optional dependencies a package has, if it has any)

Comment (AlexWaygood) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: ...I think I misunderstood what you meant — sorry. A better answer for you is that, no, the script we have in CI to install necessary stubtest requirements and then run stubtest doesn't currently know anything about `extras`. But we could probably improve that script — it's here: https://github.com/python/typeshed/blob/master/tests/stubtest_third_party.py

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/contrib/socks.pyi:19]:
```suggestion
```

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/contrib/socks.pyi:23]:
```suggestion
class SOCKSHTTPConnectionPool(HTTPConnectionPool):
    ConnectionCls: ClassVar[type[SocksConnection]]
class SOCKSHTTPSConnectionPool(HTTPSConnectionPool):
    ConnectionCls: ClassVar[type[SocksConnection]]
```

You may have to also change this line to be a `ClassVar`: https://github.com/python/typeshed/blob/df4618ef2dec1d27e363da0fc4ba8b00a2c811fe/stubs/urllib3/urllib3/connectionpool.pyi#L47

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/contrib/socks.pyi:26]:
```suggestion
    http: type[SOCKSHTTPConnectionPool]
```

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/contrib/socks.pyi:30]: This is overridden on instances, so it's wrong to annotate it as a `ClassVar`, despite the fact it has a class-level default: https://github.com/urllib3/urllib3/blob/aa3def7d242525e6e854991247c4b68583d15135/src/urllib3/contrib/socks.py#L216

```suggestion
   # has a class-level default, but is overridden on instances, so not a ClassVar
    pool_classes_by_scheme: _ConnectionPoolClasses
```

**Review by AlexWaygood — CHANGES_REQUESTED** [review]

File: stubs/urllib3/@tests/requirements-stubtest.txt
Comment (AlexWaygood) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: I like that idea, quite a lot :) though I can't speak for the other maintainers.

(A change like that should obviously be made in a separate PR, though!)

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (kkirsche) [reply stubs/urllib3/urllib3/contrib/socks.pyi:21]: https://github.com/urllib3/urllib3/blob/1.26.x/src/urllib3/connectionpool.py reviewed with a variety of updates to improve coverage of this file.

If you get stuck on any of the types, comparing the implementation of the 1.26.x branch (above) and the main 2.x branch seen [here](https://github.com/urllib3/urllib3/blob/main/src/urllib3/connectionpool.py) is a great way to verify types for some of the more trickily named parameters or attributes.

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:6]:
```suggestion
from typing import Any, ClassVar
from typing_extensions import TypeAlias
```

File: stubs/urllib3/@tests/requirements-stubtest.txt
Comment (kkirsche) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: https://github.com/python/typeshed/pull/8467 submitted

Comment (AlexWaygood) [reply stubs/urllib3/@tests/requirements-stubtest.txt:1]: Thank you!

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (srittau) [reply stubs/urllib3/urllib3/contrib/socks.pyi]: We would basically copy the signature of `HTTPConnection.__init__()`. Cumbersome, but possible.

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (AlexWaygood) [reply stubs/urllib3/urllib3/contrib/socks.pyi:21]: Thanks! That's more than I was expecting you to do in this PR, but that's fine 🙂

Incredibly, the new stubtest failure appears as though it's due to our stdlib stubs not having `abc.ABCMeta` as the metaclass for `io.IOBase` in the stdlib. I've filed #8468 to fix it.

File: stubs/urllib3/urllib3/contrib/socks.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/contrib/socks.pyi:2]:
```suggestion
from typing import ClassVar
from typing_extensions import TypedDict
```

**Comment by AlexWaygood** [comment]
The continued stubtest failure looks like a stubtest bug to me; I've filed https://github.com/python/mypy/issues/13316. We'll get a new mypy release before the end of August; hopefully either I or @hauntsaninja can fix it by then. Until then I recommend that you add an allowlist entry, with a comment linking to the stubtest bug report :)

**Comment by kkirsche** [comment]
> python/mypy#13316

Added via d8d8d0b. Hopefully I'll be able to contribute at some point to `mypy`, currently I'm not comfortable enough with the [visitor design pattern](https://refactoring.guru/design-patterns/visitor) to feel confident beyond small refactoring.

**Comment by kkirsche** [comment]
> Added via [d8d8d0b](https://github.com/python/typeshed/commit/d8d8d0b9c9bb7eec0e4683e59a53c7f16c157313). Hopefully I'll be able to contribute at some point to `mypy`, currently I'm not comfortable enough with the [visitor design pattern](https://refactoring.guru/design-patterns/visitor) to feel confident beyond small refactoring.

OK, I thought I had added it, but apparently I did it wrong? Would someone be able to help me understand what I got wrong with that commit?

**Comment by AlexWaygood** [comment]
> OK, I thought I had added it, but apparently I did it wrong? Would someone be able to help me understand what I got wrong with that commit?

You did it right, but you actually need to add two allowlist entries. Stubtest emits two errors for this class because of the fact that the class is declared in one module, but then imported and re-exported from another module!

**Comment by AlexWaygood** [comment]
> Added via [d8d8d0b](https://github.com/python/typeshed/commit/d8d8d0b9c9bb7eec0e4683e59a53c7f16c157313). Hopefully I'll be able to contribute at some point to `mypy`, currently I'm not comfortable enough with the [visitor design pattern](https://refactoring.guru/design-patterns/visitor) to feel confident beyond small refactoring.

It's a huge and complicated codebase! I've made a few contributions now, but there's still a lot that I don't really understand. Stubtest is a nice place to start contributing, though :)

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:50]: The return type here is quite important -- both mypy and pyright do quite heavy special-casing of `__exit__` methods. If they see an `__exit__` method that's annotated as returning `bool`, they'll assume it's an `__exit__` method that can sometimes swallow exceptions. But this method _can't_, because it always returns `False`, never `True`.

```suggestion
    def __exit__(self, exc_type: type[BaseException] | None, exc_val: BaseException | None, exc_tb: TracebackType | None) -> Literal[False]: ...
```

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:4]: This module imports `LifoQueue` from another submodule of `urllib3`, _not_ the stdlib `queue` module

```suggestion
from .util.queue import LifoQueue
```

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:45]:
```suggestion
    QueueCls: ClassVar[queue.Queue]
```

Where `queue` is the stdlib `queue` module

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:55]: I believe the `Union` type on the main branch is deliberate here: while `HTTPSConnectionPool`s can only have `HTTPSConnection`s, `HTTPConnectionPool`s can have `HTTPConnection`s or `HTTPSConnection`s: https://github.com/urllib3/urllib3/blob/1e985050bb498f15895478bd61acc002be83ae65/src/urllib3/connectionpool.py#L169

```suggestion
    ConnectionCls: ClassVar[type[HTTPConnection | HTTPSConnection]]
```

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:63]: If `None` is passed into the `__init__` parameter, it's set to an empty dictionary: https://github.com/urllib3/urllib3/blob/aa3def7d242525e6e854991247c4b68583d15135/src/urllib3/connectionpool.py#L202

```suggestion
    proxy_headers: Mapping[str, str]
```

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:104]: It's `str | Literal[False] | None` on the `main` branch https://github.com/urllib3/urllib3/blob/1e985050bb498f15895478bd61acc002be83ae65/src/urllib3/connectionpool.py#L942

```suggestion
    assert_hostname: str | Literal[False] | None
```

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:123]:
```suggestion
        assert_hostname: str | Literal[False] | None = ...,
```

**Review by AlexWaygood — CHANGES_REQUESTED** [review]

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi:50]: Ah, I had seen https://github.com/python/typeshed/issues/7258 and thought from the title that this was related to literal's in this sense (my bad on not reading the issue, just the title). Sorry about that. I'll fix that in a moment

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi:4]: Ah, sorry. That was a diff between their 1.26 and main branch. I had mixed up which tab was which, as it's LifoQueue from `queue` in `main`:
https://github.com/urllib3/urllib3/blob/main/src/urllib3/connectionpool.py#L3

Since `queue` stubs don't exist, I'll add them since the file is only 22 lines:
https://github.com/urllib3/urllib3/blob/1.26.x/src/urllib3/util/queue.py

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi:104]: Thanks, this was my misunderstanding on the use of Literal again — appreciate the find!

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:4]: The fact that there are two `queue` modules here is quite confusing. I think it would be more readable if we just did

```python
import queue
import urllib3.util.queue
```

And then referred to `LifoQueue` as `urllib3.util.queue.LifoQueue`, and `Queue` as `queue.Queue`. It's more verbose, but it's a lot less confusing, in my opinion.

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (Akuli) [reply stubs/urllib3/urllib3/connectionpool.pyi]: Their `import queue` imports the standard-library module named `queue`. Anything else would have to be `from urllib.util import queue` or `from .util import queue` or something similar. Here are the stubs for the standard-library `queue` module: https://github.com/python/typeshed/blob/master/stdlib/queue.pyi

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi]: That's fixed because that should have been `LifoQueue` in both locations after reviewing 

https://github.com/urllib3/urllib3/blob/1.26.x/src/urllib3/connectionpool.py

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:46]: No, stubtest was complaining because you had `ClassVar[Queue[Any]]`, when you should have had `ClassVar[type[Queue[Any]]]`

Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi]: That should be fixed in the latest revision

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi]: It's not, because (putting aside the disagreement over which queue class it should be), you're saying it's an instance of a class, but it's not, it's the class itself. Even if you want `LifoQueue`, it needs to be `ClassVar[type[LifoQueue]]`, not `ClassVar[LifoQueue]`

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi]: ^You made that update.

Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi]: Just a miscommunication caused by using async comments :)

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi]: Race conditions! Everywhere I look! More race conditions! :)

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (AlexWaygood) [inline stubs/urllib3/urllib3/connectionpool.pyi:46]: I still think this would be better as the stdlib `Queue[Any]`. It's true that at runtime, `ConnectionPool.QueueCls` is `urllib3.util.queue.LifoQueue`. But `ConnectionPool` is a base class for other `ConnectionPool` classes, and I don't think (from my impression, from reading the code) that it's mandatory for all subclasses to have their `QueueCls` attribute be a subclass of `urllib3.util.queue.LifoQueue`. It seems to me that anything with an interface similar to the stdlib `queue.Queue` class will do. There's evidence for this in the fact that they scrapped the `urllib3.util.queue` submodule in the `main` branch and switched to using the stdlib `queue` module.

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi:46]: Anyway, this is minor, other than this it looks good to me now :)

**Review by AlexWaygood — APPROVED** [review]

File: stubs/urllib3/urllib3/connectionpool.pyi
Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi:46]: 👍  I think I got mixed up between the `type[_QueueT]` change and the `Queue[Any]` change. I agree with you on that.

If you are comfortable with it, I will update this to: `ClassVar[type[Queue[Any]]]`

and change the imports to be fully qualified to differentiate between the two different queues.

Sound good?

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi:46]: That sounds good to me! 👍

Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi]: That was changed in 3e231f4 (I expect a pre-commit will trigger). The only thing I may have missed was if there was a better way to import the `LifoQueue` to avoid the overlapping module names. As it stands, I've updated it to `import queue` for the standard library instance and usage as `queue.Queue[Any]` and have left the urllib3 `LifoQueue` as `from .util.queue import LifoQueue`.

Hopefully thats correct

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi]: How about `from .util import queue as urllib3queue`? And then refer to it as `urllib3queue.LifoQueue`

Comment (kkirsche) [reply stubs/urllib3/urllib3/connectionpool.pyi]: I had thought that was going to trigger a stubtest warning about it not existing at runtime, but I could just be getting confused about what is checked and what isn't. 

That's a good approach, thank you for the suggestion, worst case we can do it as `from .util import queue as _urllib3queue` to avoid the check

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi]: All imports in stub files are considered "private to the stub" (not re-exported), _unless_ they fall into one of the two categories:

- `import foo as foo` -- the `as` has to be truly redundant; if you do `import foo as bar`, it's still considered private, but `import foo as foo` is considered a public re-export.
- `from foo import *`

Comment (AlexWaygood) [reply stubs/urllib3/urllib3/connectionpool.pyi]: -- A third exception is if a name is included in `__all__`, it's always considered re-exported, even if it's imported via a means that would usually make it a "private" import.

**Review by AlexWaygood — APPROVED** [review]

**Review by srittau — APPROVED** [review]
Not an in-depth review, but everything looked fine at first glance. I'll defer to @AlexWaygood's judgement.

**Comment by AlexWaygood** [comment]
Let's land it. Thanks for working on this @kkirsche, this was a tricky one in lots of ways!

**Comment by kkirsche** [comment]
My pleasure, trying to work towards larger and loftier goals, so hopefully experiences like this will help

**Pull request merged**
